### PR TITLE
feat: add search engine reindexing cli task

### LINKED ISSUE #4335: [FEATURE] Add new argilla cli option to reindex all entities into search engine
**Is your feature request related to a problem? Please describe.**

Sometimes our search engine mapping is updated/changed and we don't have an easy way to allow a general reindex of previous created data on our database.

For example after a new version of Argilla is released and the search engine mapping is changed for some entities, users don't have an easy way to reindex all the old data to have support for searching with the new mappings.

**Describe the solution you'd like**

A possible solution could be to create an Argilla cli option that reindex all the Argilla entities into the search engine.

Something similar to the following:

```sh
$ argilla server reindex
reindexing datasets...
reindexing records...
reindexing responses...
# Will reindex datasets, records, responses, suggestions, and the rest of entities etc.
```

Optionally we should show some way of progress or at least some messages showing which entities are reindexed.

Once that we have this Argilla cli option we can use it on our docker images so every time that the server is up we reindex or if that's too much we can add notes to the release saying that a manual reindex is necessary (in the case that we don't want to automatically do it).

### PR DESCRIPTION
**Pull request by jfcalvo**
# Description

This PR adds a new cli task to reindex datasets and records so we can use it once ElasticSearch/OpenSearch mappings are updated.

To execute the cli task use the following command:

```sh
$ argilla server search_engine reindex
```

This task will iterate over all the datasets in the database, reindexing them and doing the same for all records for each dataset.

> [!NOTE]
> We are using server-side cursors (streams) for fetch the collection of datasets and records in the cli task. For more information take a look to [SQLAlchemy documentation](https://docs.sqlalchemy.org/en/20/core/connections.html#engine-stream-results).

Closes #4335 

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [x] Tested manually.
- [x] Tested manually using PostgreSQL.

**Checklist**

- [ ] I added relevant documentation
- [ ] follows the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/cli/server/search_engine/reindex.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        return (await db.execute(select(func.count(Dataset.id)))).scalar()
+
+    @classmethod
+    async def count_dataset_records(cls, db: AsyncSession, dataset: Dataset) -> int:
+        return (await db.execute(select(func.count(Record.id)).filter_by(dataset_id=dataset.id))).scalar()
+
+
+async def _reindex_datasets(db: AsyncSession, search_engine: SearchEngine, progress: Progress) -> None:
+    task = progress.add_task(
+        f"reindexing datasets...",
```
Comment (frascuchon) [inline src/argilla/cli/server/search_engine/reindex.py]: Here you can specify that only the Feedback datasets will be reindexed. Just to clarify to users

File: src/argilla/cli/server/search_engine/reindex.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                selectinload(Dataset.questions),
+                selectinload(Dataset.metadata_properties),
+                selectinload(Dataset.vectors_settings),
+            )
+            .execution_options(yield_per=cls.YIELD_PER)
+        )
+
+        async for dataset in stream.scalars():
+            await search_engine.delete_index(dataset)
+            await search_engine.create_index(dataset)
```
Comment (frascuchon) [inline src/argilla/cli/server/search_engine/reindex.py:71]: For "soft" changes we can use the update mappings and the update settings endpoints:

See docs:
- https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html
- https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-update-settings.html

If we take care of mapping changes and we do not modify the field type, we could always update mappings without deleting the index.  See [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html#updating-field-mappings)

File: tests/unit/cli/server/search_engine/test_reindex.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import pytest
+from typer import Typer
+from typer.testing import CliRunner
+
+
+@pytest.mark.asyncio
```
Comment (frascuchon) [inline tests/unit/cli/server/search_engine/test_reindex.py]:
```suggestion
```

File: src/argilla/cli/server/search_engine/reindex.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+async def _reindex_datasets(db: AsyncSession, search_engine: SearchEngine, progress: Progress) -> None:
+    task = progress.add_task(
+        f"reindexing datasets...",
+        total=await Reindexer.count_datasets(db),
+    )
+
+    async for dataset in Reindexer.reindex_datasets(db, search_engine):
+        await _reindex_dataset_records(db, search_engine, progress, dataset)
+
+        progress.advance(task)
```
Comment (gabrielmbmb) [inline src/argilla/cli/server/search_engine/reindex.py:133]: is this `progress.advance` needed?

Comment (jfcalvo) [reply src/argilla/cli/server/search_engine/reindex.py:133]: Yes, I think so. We will have two progress bars one for datasets and another one for records (per dataset).

File: src/argilla/cli/server/search_engine/reindex.py (hunk 1 above)
Comment (jfcalvo) [reply src/argilla/cli/server/search_engine/reindex.py]: Done.

File: tests/unit/cli/server/search_engine/test_reindex.py (hunk 3 above)
Comment (jfcalvo) [reply tests/unit/cli/server/search_engine/test_reindex.py]: Done.

File: src/argilla/cli/server/search_engine/reindex.py (hunk 4 above)
Comment (gabrielmbmb) [reply src/argilla/cli/server/search_engine/reindex.py:133]: ahh, got it!

**Comment by frascuchon** [comment]
It would be nice to add some references in docs about this new task. I think the better place should be [here](https://docs.argilla.io/en/latest/getting_started/installation/configurations/database_migrations.html#reindex-a-dataset)

**Comment by jfcalvo** [comment]
> It would be nice to add some references in docs about this new task. I think the better place should be [here](https://docs.argilla.io/en/latest/getting_started/installation/configurations/database_migrations.html#reindex-a-dataset)

Ok, I will add a new section in the docs about this task.

**Comment by jfcalvo** [comment]
> > It would be nice to add some references in docs about this new task. I think the better place should be [here](https://docs.argilla.io/en/latest/getting_started/installation/configurations/database_migrations.html#reindex-a-dataset)
> 
> Ok, I will add a new section in the docs about this task.

Done.

**Review by frascuchon — APPROVED** [review]

**Pull request merged**